records/2022-12-13.upgrade-database.md

```diff
@@ -0,0 +1,104 @@
+
+### Option B: Snapshot approach
+In this approach we would make a copy of the database and upgrade that copy, then switch the app to use the new database. The benefit is that if something doesn't work we can quickly switch back to the old database. This approach still requires downtime during the process though, since data added to the old database would not be in the new database.
+
+### Option C: AWS DMS approach
+We can use the AWS Database Migration Service. Here we create a new database with the new version, then turn on DMS. The old data will be copied into the new database. Also, any data added during this time will also be copied over. We then stop Notify, let any remaining data go to the new database, point Notify at the new database, and restart.
+This is a fairly straightforward approach (as these things go) and will have less downtime than other methods. Note that DMS does not work with all data types or database configurations, so it might not be an option.
+
+### Option D: Replication-based approach using pglogical
+This approach use the `pglogical` database extension along with database replication sets (A replication set is way to mark a collection of tables to be replicated.). In many ways this is very similar to Option C, however in this case once the new database is populated it stays up to date faster, and there is no "wait for last entries to transfer" downtime. Rather we just have to turn off Notify, switch to the new database, and turn Notify back on (and then turn off the extension). There are restriction on what database configurations are acceptible, however, and this is a more manual process than using DMS.
+
+### Discussion
+
+We want to minimize downtime and maximize probability of success. 
+- There probably isn't too much to be gained performance-wise by just moving to 11.16, so we should move to 12.12 or 13.14. Just moving to 12.12 will increase the probability of success, at the expense of having to do another migration at a future point.
+- Note that one benefit to just a minor upgrade would be that we could be confident that any issues that arise are due to our migration and not to the version change. So we could use a 11.16 upgrade as an opportunity to gain familiarity with the migration process.
+- Option A is probably the worst on both scores. 
+- Option B give us the ability to quickly switch to the old database if things go wrong, but the downtime will be the same (likely a bit longer because we need to also take the snapshot). 
+- Options C and D probably have the same downtime for us (if we do this on a Saturday evening, for example, there will not be may notifications sent during Option C's "wait for last entries to transfer" window)
+- Option C may be less manual and more documented / supported. Both approaches require multiple manual steps however.
```

Comment:
option c seems the most straightfoward

Reply:
I'd vote for option C as well since it seems like AWS would take care of most of the complexity for us.

Reply:
It seems the best. What is the required effort with this option?

Reply:
Honestly not sure as I haven't done an upgrade of this scale before.  We'll have to do more research and test it out with our Staging upgrade.

Reply:
A gross estimate could work, relative to other options (not time related).

Reply:
Good reference @sastels , it gives a better perspective. We can include it and compare to other commands listing?